cpp/include/raft/spatial/knn/detail/ivf_pq_search.cuh

```diff
@@ -0,0 +1,1203 @@
+          we extend the dimension K of the GEMM to compute it together with all the dot products:
+
+          `cq_distances[i, j] = 0.5 |luster_centers[j]|^2 - (queries[i], cluster_centers[j])`
+
+          This is a monotonous mapping of the proper L2 distance.
+
+        IP distance:
+          `cq_distances[i, j] = - (queries[i], cluster_centers[j])`
+
+          This is a negative inner-product distance. We minimize it to find the similar clusters.
+
+          NB: cq_distances is NOT used further in ivfpq_search.
+     */
+    float norm_factor;
+    switch (index.metric()) {
+      case raft::distance::DistanceType::L2Expanded: norm_factor = 1.0 / -2.0; break;
+      case raft::distance::DistanceType::InnerProduct: norm_factor = 0.0; break;
+      default: RAFT_FAIL("Unsupported distance type %d.", int(index.metric()));
+    }
+    utils::copy_fill(queries_batch,
+                     index.dim(),
+                     queries + static_cast<size_t>(index.dim()) * i,
+                     index.dim(),
+                     cur_queries.data(),
+                     index.dim_ext(),
+                     norm_factor,
+                     stream);
+
+    float alpha;
+    float beta;
+    uint32_t gemm_k = index.dim();
+    switch (index.metric()) {
+      case raft::distance::DistanceType::L2Expanded: {
+        alpha  = -2.0;
+        beta   = 0.0;
+        gemm_k = index.dim() + 1;
+        RAFT_EXPECTS(gemm_k <= index.dim_ext(), "unexpected gemm_k or dim_ext");
+      } break;
+      case raft::distance::DistanceType::InnerProduct: {
+        alpha = -1.0;
+        beta  = 0.0;
+      } break;
+      default: RAFT_FAIL("Unsupported distance type %d.", int(index.metric()));
+    }
+    linalg::gemm(handle,
+                 true,
+                 false,
+                 index.n_lists(),
+                 queries_batch,
+                 gemm_k,
+                 &alpha,
+                 index.centers().data_handle(),
+                 index.dim_ext(),
+                 cur_queries.data(),
+                 index.dim_ext(),
+                 &beta,
+                 qc_distances.data(),
+                 index.n_lists(),
+                 stream);
+
+    // Rotate queries
+    alpha = 1.0;
+    beta  = 0.0;
+    linalg::gemm(handle,
+                 true,
+                 false,
+                 index.rot_dim(),
+                 queries_batch,
+                 index.dim(),
+                 &alpha,
+                 index.rotation_matrix().data_handle(),
+                 index.dim(),
+                 cur_queries.data(),
+                 index.dim_ext(),
+                 &beta,
+                 rot_queries.data(),
+                 index.rot_dim(),
+                 stream);
+
+    {
+      // Select neighbor clusters for each query.
+      rmm::device_uvector<float> cluster_dists(max_queries * params.n_probes, stream, mr);
+      select_topk<float, uint32_t>(qc_distances.data(),
```

Comment:
Please factor the `select_topk` op together with the corresponding `gemm` into a separate function.